File: python/src/aumos_sdk/_retry.py
"""Exponential backoff retry logic for the AumOS Python SDK.

Implements the "full jitter" strategy from AWS architecture blog:
delay = random(0, min(cap, base * 2^attempt))

Usage:
    result = await with_retry(lambda: client._http.get("/agents"))
"""

from __future__ import annotations

import asyncio
import random
from collections.abc import Callable
from typing import TypeVar

import httpx

T = TypeVar("T")

RETRYABLE_STATUS_CODES: frozenset[int] = frozenset({429, 500, 502, 503, 504})
DEFAULT_MAX_RETRIES: int = 3
DEFAULT_BASE_DELAY_SECONDS: float = 0.5
DEFAULT_MAX_DELAY_SECONDS: float = 60.0


async def with_retry(
    func: Callable[[], T],
    max_retries: int = DEFAULT_MAX_RETRIES,
    base_delay: float = DEFAULT_BASE_DELAY_SECONDS,
    max_delay: float = DEFAULT_MAX_DELAY_SECONDS,
) -> T:
    """Execute an async callable with exponential backoff retry on transient failures.

    Uses full jitter strategy: delay = random(0, min(cap, base * 2^attempt)).
    Respects the Retry-After header on 429 responses.

    Args:
        func: Async callable to execute and retry on transient failure.
        max_retries: Maximum number of retry attempts after the initial attempt (default 3).
        base_delay: Base delay in seconds for the first retry (default 0.5).
        max_delay: Maximum delay cap in seconds (default 60.0).

    Returns:
        The return value of func on success.

    Raises:
        httpx.HTTPStatusError: On non-retryable HTTP errors.
        AumOSError: After max_retries exhausted.
    """
    last_error: Exception | None = None

    for attempt in range(max_retries + 1):
        try:
            return await func()  # type: ignore[return-value]
        except httpx.HTTPStatusError as exc:
            if exc.response.status_code not in RETRYABLE_STATUS_CODES:
                raise
            last_error = exc
            if exc.response.status_code == 429:
                retry_after_header = exc.response.headers.get("Retry-After")
                if retry_after_header is not None:
                    delay = min(float(retry_after_header), max_delay)
                else:
                    cap = min(max_delay, base_delay * (2**attempt))
                    delay = random.uniform(0, cap)
            else:
                cap = min(max_delay, base_delay * (2**attempt))
                delay = random.uniform(0, cap)
        except (httpx.ConnectError, httpx.TimeoutException) as exc:
            last_error = exc
            cap = min(max_delay, base_delay * (2**attempt))
            delay = random.uniform(0, cap)

        if attempt < max_retries:
            await asyncio.sleep(delay)

    raise RuntimeError(
        f"AumOS request failed after {max_retries} retries: {last_error}"
    ) from last_error

File: typescript/src/_retry.ts
/**
 * Exponential backoff retry logic for the AumOS TypeScript SDK.
 *
 * Uses the "full jitter" strategy: delay = random(0, min(cap, base * 2^attempt)).
 * Respects Retry-After headers on 429 responses.
 *
 * @example
 * const result = await withRetry(() => fetch(url))
 */

export interface RetryConfig {
  /** Maximum number of retry attempts after the initial attempt. Default: 3 */
  maxRetries: number;
  /** Base delay in milliseconds for the first retry. Default: 500 */
  baseDelayMs: number;
  /** Maximum delay cap in milliseconds. Default: 60_000 */
  maxDelayMs: number;
}

export const DEFAULT_RETRY_CONFIG: RetryConfig = {
  maxRetries: 3,
  baseDelayMs: 500,
  maxDelayMs: 60_000,
};

const RETRYABLE_STATUS_CODES = new Set([429, 500, 502, 503, 504]);

interface RetryableError extends Error {
  status?: number;
  headers?: Headers | Record<string, string>;
}

function isRetryableError(err: unknown): boolean {
  if (!(err instanceof Error)) return false;
  const retryable = err as RetryableError;
  if (retryable.status !== undefined) {
    return RETRYABLE_STATUS_CODES.has(retryable.status);
  }
  // Network-level errors (no status code) are always retryable
  return (
    err.message.includes("ECONNREFUSED") ||
    err.message.includes("ETIMEDOUT") ||
    err.message.includes("fetch failed")
  );
}

function getRetryAfterMs(err: unknown): number | null {
  if (!(err instanceof Error)) return null;
  const retryable = err as RetryableError;
  if (retryable.status !== 429) return null;

  const headers = retryable.headers;
  if (!headers) return null;

  const retryAfter =
    typeof headers === "object" && "get" in headers
      ? (headers as Headers).get("retry-after")
      : (headers as Record<string, string>)["retry-after"];

  if (retryAfter === null || retryAfter === undefined) return null;
  const parsed = parseFloat(retryAfter);
  return isNaN(parsed) ? null : parsed * 1000;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Execute an async function with automatic exponential backoff retry.
 *
 * @param fn - Async function to execute and retry on transient failure.
 * @param config - Retry configuration (uses DEFAULT_RETRY_CONFIG if omitted).
 * @returns The resolved value of fn on success.
 * @throws The last error after all retry attempts are exhausted.
 */
export async function withRetry<T>(
  fn: () => Promise<T>,
  config: RetryConfig = DEFAULT_RETRY_CONFIG,
): Promise<T> {
  let lastError: unknown;

  for (let attempt = 0; attempt <= config.maxRetries; attempt++) {
    try {
      return await fn();
    } catch (err) {
      lastError = err;

      if (!isRetryableError(err)) {
        throw err;
      }

      if (attempt === config.maxRetries) {
        break;
      }

      const retryAfterMs = getRetryAfterMs(err);
      let delayMs: number;

      if (retryAfterMs !== null) {
        delayMs = Math.min(retryAfterMs, config.maxDelayMs);
      } else {
        const cap = Math.min(config.maxDelayMs, config.baseDelayMs * 2 ** attempt);
        delayMs = Math.random() * cap;
      }

      await sleep(delayMs);
    }
  }

  throw new Error(
    `AumOS request failed after ${config.maxRetries} retries: ${String(lastError)}`,
    { cause: lastError },
  );
}
